ow_power_system/src/PowerSystemPack.cpp

```diff
@@ -0,0 +1,375 @@
+    sort(samplesRUL.begin(), samplesRUL.end());
+    double eod_median = samplesRUL.at(samplesRUL.size() / 2);
+    auto now = MessageClock::now();
+    auto now_s = duration_cast<std::chrono::seconds>(now.time_since_epoch());
+    double rul_median = eod_median - now_s.count();
+    //rul_msg.data = rul_median;
+
+    // Determine the median SOC.
+    UData currentSOC = eod_event.getState()[0];
+    auto samplesSOC = currentSOC.getVec();
+    sort(samplesSOC.begin(), samplesSOC.end());
+    double soc_median = samplesSOC.at(samplesSOC.size() / 2);
+    //soc_msg.data = soc_median;
+
+    // Determine the Battery Temperature
+    auto stateSamples = eod_event.getSystemState()[0];
+    std::vector<double> state;
+    for (auto sample : stateSamples)
+    {
+      state.push_back(sample[0]);
+    }
+
+    // HACK ALERT:
+    // The asynchronous prognoser does not have a function to get its model like
+    // the simple prognoser does. As such, one can instantiate a simple prognoser
+    // and get its model; it should be identical. However, a getter function would
+    // be much more ideal and more efficient.
+    // Construct a simple prognoser to get the battery model.
+    auto prognoser_config_path = ros::package::getPath("ow_power_system") + "/config/prognoser.cfg";
+    ConfigMap prognoser_config(prognoser_config_path);
+    std::unique_ptr<PCOE::Prognoser> temp_prog =
+      PrognoserFactory::instance().Create("ModelBasedPrognoser", prognoser_config);
+
+    auto& model = dynamic_cast<ModelBasedPrognoser*>(temp_prog.get())->getModel();
+
+    auto model_output = model.outputEqn(now_s.count(), static_cast<PrognosticsModel::state_type>(state));
+
+    // Store the newly obtained data.
+    EoD_events[node_number][0] = rul_median;
+    EoD_events[node_number][1] = soc_median;
+    EoD_events[node_number][2] = model_output[TEMPERATURE_INDEX];
+  }
+
+private:
+  MessageBus& bus;
+  std::string identifier = "Node X";
+  int node_number = -1;
+};
+
+PowerSystemPack::PowerSystemPack()
+{ }
+
+void PowerSystemPack::InitAndRun()
+{
+  if (!initNodes())
+  {
+    ROS_ERROR_STREAM("Power system pack failed to initialize nodes");
+    return;
+  }
+
+  if (!initTopics())
+  {
+    ROS_ERROR_STREAM("Power system pack failed to initialize topics");
+    return;
+  }
+
+  // Initialize EoD_events and previous_times.
+  for (int i = 0; i < NUM_NODES; i++)
+  {
+    previous_times[i] = 0;
+    for (int j = 0 ; j < 3; j++)
+    {
+      EoD_events[i][j] = -1;
+    }
+  }
+
+  // Construct the prediction handlers.
+  PredictionHandler handler_0(bus[0], node_names[0]);
+  PredictionHandler handler_1(bus[1], node_names[1]);
+  PredictionHandler handler_2(bus[2], node_names[2]);
+  PredictionHandler handler_3(bus[3], node_names[3]);
+  PredictionHandler handler_4(bus[4], node_names[4]);
+  PredictionHandler handler_5(bus[5], node_names[5]);
+  PredictionHandler handler_6(bus[6], node_names[6]);
+  PredictionHandler handler_7(bus[7], node_names[7]);
```

Comment:
I finally figured out how to declare both `PredictionHandlers` and `AsyncPrognosers` within their own loops, so now the code properly supports multiple nodes and all it takes is editing the `NUM_NODES` constant. While it would still be better to add `PredictionHandler`s and `AsyncPrognoser`s to another class, I'm going to leave it as-is and it can be revisited in a future PR.